### PR TITLE
Issue #178: Fix malformed HQL when @additional_filters@ is preceded by and

### PR DESCRIPTION
**Pull request by RomanMagnoli**
## Summary

Fixes a `QuerySyntaxException` thrown when creating records via the EtendoRX Datasource API when the selector HQL contains `and @additional_filters@` and the filter expression evaluates to a non-empty string.

The `else` branch in `buildHQLQuery()` was only replacing the bare `@additional_filters@` placeholder, leaving the preceding `and` keyword in place and producing malformed HQL like `... and AND c.id = '102' ...`.

The fix applies the same two-step replacement already used in the empty-filter branch:

```java
result = hqlQuery.replace("and @additional_filters@", additionalFilters)
    .replace("@additional_filters@", additionalFilters);
```

## Changes

- `SelectorHandlerUtil.java` — fix two-step replacement in `buildHQLQuery()` else branch
- `SelectorHandlerUtilTest.java` — add 3 regression tests covering empty filters, `and`-preceded placeholder, and standalone placeholder

## Jira

ETP-3623

### CODE REVIEW

**Review by etendobot — APPROVED** [review]

File: src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilTest.java [hunk 1]
```diff
@@ -51,6 +51,8 @@ public class SelectorHandlerUtilTest extends WeldBaseTest {
     private static final String FALLBACK_NAME_NORMALIZED = "fallback$name";
     private static final String INP_C_BPARTNER_ID = "inpcBpartnerId";
     private static final String INP_C_BPARTNER_ID_DES = "inpcBpartnerId_DES";
+    private static final String INP_M_PRODUCT_ID = INP_M_PRODUCT_ID;
+    private static final String SALES_PRICE_LIST_FILTER = SALES_PRICE_LIST_FILTER;
```
Comment (etendobot) [inline src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilTest.java]: **[RESOLVED]**  > [!WARNING]
**⚠️ Blocking Issue**

Assign the string literal "inpmProductId" to the constant instead of referencing the constant itself.

```java
private static final String INP_M_PRODUCT_ID = "inpmProductId";
```

**Rationale**: Recursive initialization (assigning a variable to itself during declaration) is a compilation error in Java. The literal value "inpmProductId" was previously used and should be assigned to the constant.

File: src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilTest.java [hunk 2]
```diff
@@ -51,6 +51,8 @@ public class SelectorHandlerUtilTest extends WeldBaseTest {
...
```
Comment (etendobot) [inline src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilTest.java]: **[RESOLVED]**  > [!WARNING]
**⚠️ Blocking Issue**

Assign the string literal "pl.salesPriceList = true" to the constant instead of referencing the constant itself.

```java
private static final String SALES_PRICE_LIST_FILTER = "pl.salesPriceList = true";
```

**Rationale**: Recursive initialization (assigning a variable to itself during declaration) is a compilation error in Java. The filter clause "pl.salesPriceList = true" was previously used and should be assigned to the constant.

**Review by etendobot — CHANGES_REQUESTED** [review]
Changes requested by agent. Please resolve blocking issues.

File: src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilBaseTest.java [hunk 3]
```diff
@@ -0,0 +1,170 @@
+package com.etendoerp.etendorx.utils;
+
```
Comment (etendobot) [inline src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilBaseTest.java:18]: **[RESOLVED]**  > [!WARNING]
**⚠️ Blocking Issue**

Add the mandatory Etendo license header at the top of the file.

```java
/*\n *************************************************************************\n * The contents of this file are subject to the Etendo License\n * (the "License"), you may not use this file except in compliance\n * with the License.\n * You may obtain a copy of the License at\n * https://github.com/etendosoftware/etendo_core/blob/main/legal/Etendo_license.txt\n * Software distributed under the License is distributed on an\n * "AS IS" basis, WITHOUT WARRANTY OF ANY KIND, either express or\n * implied. See the License for the specific language governing rights\n * and limitations under the License.\n * All portions are Copyright (C) 2021-2024 FUTIT SERVICES, S.L\n * All Rights Reserved.\n * Contributor(s): Futit Services S.L.\n *************************************************************************\n */\npackage com.etendoerp.etendorx.utils;
```

**Rationale**: New Java source files in the com.etendoerp.* namespace must include the standard Etendo license header at the very top.

File: src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilBaseTest.java [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    String callFullfillSessionsVariables(String whereClause, Map<String, String> db2Input,
+            JSONObject dataInpFormat) throws ReflectiveOperationException {
+        Method method = SelectorHandlerUtil.class.getDeclaredMethod("fullfillSessionsVariables",
+                String.class, Map.class, JSONObject.class);
+        method.setAccessible(true);
+        return (String) method.invoke(null, whereClause, db2Input, dataInpFormat);
+    }
+
+    HashMap<String, String> callConvertToHashMAp(JSONObject dataInpFormat) throws ReflectiveOperationException {
+        Method method = SelectorHandlerUtil.class.getDeclaredMethod("convertToHashMAp", JSONObject.class);
```
Comment (etendobot) [inline src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilBaseTest.java:109]: **Suggestion**

Rename the method callConvertToHashMAp to callConvertToHashMap to follow standard naming conventions.

```java
    HashMap<String, String> callConvertToHashMap(JSONObject dataInpFormat) throws ReflectiveOperationException {
```

**Rationale**: CamelCase should be used consistently; "MAp" should be "Map".

File: src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilBaseTest.java [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+        method.setAccessible(true);
```
Comment (etendobot) [inline src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilBaseTest.java:110]: **Suggestion**

Verify the method name in SelectorHandlerUtil and ensure the reflection lookup uses the correct name (likely convertToHashMap).

```java
        Method method = SelectorHandlerUtil.class.getDeclaredMethod("convertToHashMap", JSONObject.class);
```

**Rationale**: The method name being looked up via reflection likely contains the same typo as the helper method name. If the source class 'SelectorHandlerUtil' has the correct name, this will cause a NoSuchMethodException.

File: src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilHQLTest.java [hunk 6]
```diff
@@ -0,0 +1,462 @@
+package com.etendoerp.etendorx.utils;
+
```
Comment (etendobot) [inline src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilHQLTest.java:18]: **[RESOLVED]**  > [!WARNING]
**⚠️ Blocking Issue**

Add the mandatory Etendo License header at the top of the file.

```java
/*\n *************************************************************************\n * The contents of this file are subject to the Etendo License\n * (the \"License\"), you may not use this file except in compliance\n * with the License.\n * You may obtain a copy of the License at\n * https://github.com/etendosoftware/etendo_core/blob/main/legal/Etendo_license.txt\n * Software distributed under the License is distributed on an\n * \"AS IS\" basis, WITHOUT WARRANTY OF ANY KIND, either express or\n * implied. See the License for the specific language governing rights\n * and limitations under the License.\n * All portions are Copyright (C) 2021-2026 FUTIT SERVICES, S.L\n * All Rights Reserved.\n * Contributor(s): Futit Services S.L.\n *************************************************************************\n */\npackage com.etendoerp.etendorx.utils;
```

**Rationale**: Every new Java class in a com.etendoerp.* module must include the standard Etendo license header as the first block comment in the file to comply with Etendo development standards and legal requirements.

**Review by etendobot — CHANGES_REQUESTED** [review]
Changes requested by agent. Please resolve blocking issues.

File: src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilBaseTest.java (hunk 3 above)
Comment (etendobot) [inline src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilBaseTest.java:18]: **[RESOLVED]**  > [!WARNING]
**⚠️ Blocking Issue**

Add the mandatory Etendo license header at the very top of the file before the package declaration.

```java
/*
 *************************************************************************
 * The contents of this file are subject to the Etendo License
 * (the "License"), you may not use this file except in compliance
 * with the License.
 * You may obtain a copy of the License at
 * https://github.com/etendosoftware/etendo_core/blob/main/legal/Etendo_license.txt
 * Software distributed under the License is distributed on an
 * "AS IS" basis, WITHOUT WARRANTY OF ANY KIND, either express or
 * implied. See the License for the specific language governing rights
 * and limitations under the License.
 * All portions are Copyright (C) 2021-2026 FUTIT SERVICES, S.L
 * All Rights Reserved.
 * Contributor(s): Futit Services S.L.
 *************************************************************************
 */
package com.etendoerp.etendorx.utils;
```

**Rationale**: New Java source files in the com.etendoerp.* namespace must include the standard Etendo license header as the first block comment in the file to ensure legal compliance and project standards.

File: src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilHQLTest.java (hunk 6 above)
Comment (etendobot) [inline src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilHQLTest.java:18]: **[RESOLVED]**  > [!WARNING]
**⚠️ Blocking Issue**

Add the mandatory Etendo License header at the top of the file.

```java
/*\n *************************************************************************\n * The contents of this file are subject to the Etendo License\n * (the \"License\"), you may not use this file except in compliance\n * with the License.\n * You may obtain a copy of the License at\n * https://github.com/etendosoftware/etendo_core/blob/main/legal/Etendo_license.txt\n * Software distributed under the License is distributed on an\n * \"AS IS\" basis, WITHOUT WARRANTY OF ANY KIND, either express or\n * implied. See the License for the specific language governing rights\n * and limitations under the License.\n * All portions are Copyright (C) 2021-2026 FUTIT SERVICES, S.L\n * All Rights Reserved.\n * Contributor(s): Futit Services S.L.\n *************************************************************************\n */\npackage com.etendoerp.etendorx.utils;
```

**Rationale**: Every new Java class in a com.etendoerp.* module must include the standard Etendo license header as the first block comment in the file to comply with development standards and legal requirements.

File: src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilTest.java [hunk 7]
```diff
@@ -3,172 +3,30 @@
...
-    }
-
-    private String callAddFilterClause(Selector selector, HashMap<String, String> hs1,
-            HttpServletRequest request) throws ReflectiveOperationException {
-        Method method = SelectorHandlerUtil.class.getDeclaredMethod("addFilterClause",
-                Selector.class, HashMap.class, HttpServletRequest.class);
-        method.setAccessible(true);
-        return (String) method.invoke(null, selector, hs1, request);
-    }
+public class SelectorHandlerUtilTest extends SelectorHandlerUtilBaseTest {
```
Comment (etendobot) [inline src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilTest.java:30]: **[RESOLVED]**  > [!WARNING]
**⚠️ Blocking Issue**

Verify that the tests for 'buildHQLQuery', especially 'testBuildHQLQuery_NonEmptyFilters_AndPreceded_DoesNotDuplicateAND', have been correctly migrated to 'SelectorHandlerUtilBaseTest'.

```java
public class SelectorHandlerUtilTest extends SelectorHandlerUtilBaseTest { // Ensure buildHQLQuery tests are present in the base class
```

**Rationale**: The PR title mentions a fix for malformed HQL when '@additional_filters@' is preceded by 'and'. The tests for this logic (buildHQLQuery) were previously in this file but are now being deleted. If these tests are not migrated to SelectorHandlerUtilBaseTest or another test class, the PR fails to verify the fix and introduces a regression in test coverage.

File: src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilTest.java [hunk 8]
```diff
@@ -3,172 +3,30 @@
...
 import org.openbravo.model.ad.datamodel.Column;
-import org.openbravo.model.ad.ui.Field;
-import org.openbravo.model.ad.ui.Tab;
-import org.openbravo.userinterface.selector.Selector;
 import org.openbravo.userinterface.selector.SelectorField;
 
 /**
  * Unit tests for SelectorHandlerUtil class.
+ * Covers: processValue, convertRowToJSONObject, fullfillSessionsVariables,
+ * convertToHashMAp, getNormalizedFieldName, getTargetKey.
```
Comment (etendobot) [inline src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilTest.java]: **Suggestion**

Consider correcting 'fullfill' to 'fulfill' in the Javadoc (and the method name if possible).

```java
+ * Covers: processValue, convertRowToJSONObject, fulfillSessionsVariables,
```

**Rationale**: The word 'fullfill' is a common misspelling of 'fulfill'. Even if the underlying method name contains the typo, the Javadoc should ideally use the correct spelling or the code should be refactored to fix the typo.

File: src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilTest.java [hunk 9]
```diff
@@ -3,172 +3,30 @@
...
  * All tests invoke actual production code via reflection to ensure SonarQube coverage.
```
Comment (etendobot) [inline src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilTest.java:27]: **Suggestion**

Correct the casing of 'HashMAp' to 'HashMap' in the Javadoc.

```java
+ * convertToHashMap, getNormalizedFieldName, getTargetKey.
```

**Rationale**: The word 'HashMAp' uses inconsistent casing (capital 'MA'). It should follow standard Java naming conventions (HashMap).

**Review by etendobot — CHANGES_REQUESTED** [review]
Changes requested by agent. Please resolve blocking issues.

File: src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilTest.java (hunk 7 above)
Comment (RomanMagnoli) [reply src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilTest.java:30]: /false-positive

File: src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilTest.java [hunk 10]
```diff
@@ -376,15 +234,15 @@ public void testConvertToHashMAp_BasicConversion() throws ReflectiveOperationExc
     @Test
     public void testConvertToHashMAp_EmptyJson() throws ReflectiveOperationException {
         JSONObject json = new JSONObject();
```
Comment (etendobot) [inline src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilTest.java:236]: **Suggestion**

Rename the test method to 'testConvertToHashMap_BasicConversion' to fix the typo and match the updated helper method name.

```java
public void testConvertToHashMap_BasicConversion() throws ReflectiveOperationException {
```

**Rationale**: The method name contains a typo ('HashMAp'). Since the internal calls were updated to 'callConvertToHashMap', the test method name should also be updated for consistency and clarity.

Comment (etendobot) [inline src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilTest.java:236]: **Suggestion**

Rename the test method to 'testConvertToHashMap_EmptyJson'.

```java
public void testConvertToHashMap_EmptyJson() throws ReflectiveOperationException {
```

**Rationale**: The method name contains a typo ('HashMAp'). Updating it ensures consistency with the helper method 'callConvertToHashMap'.

File: src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilTest.java [hunk 11]
```diff
@@ -376,15 +234,15 @@ public void testConvertToHashMAp_BasicConversion() throws ReflectiveOperationExc
...
-        HashMap<String, String> result = callConvertToHashMAp(json);
+        HashMap<String, String> result = callConvertToHashMap(json);
         assertTrue(result.isEmpty());
     }
 
     @Test
     public void testConvertToHashMAp_BooleanValue() throws ReflectiveOperationException, JSONException {
         JSONObject json = new JSONObject();
```
Comment (etendobot) [inline src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilTest.java:243]: **Suggestion**

Rename the test method to 'testConvertToHashMap_BooleanValue'.

```java
public void testConvertToHashMap_BooleanValue() throws ReflectiveOperationException, JSONException {
```

**Rationale**: The method name contains a typo ('HashMAp'). Updating it ensures consistency with the helper method 'callConvertToHashMap'.

**Review by etendobot — CHANGES_REQUESTED** [review]
Changes requested by agent. Please resolve blocking issues.

File: src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilBaseTest.java [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+    String callFullfillSessionsVariables(String whereClause, Map<String, String> db2Input,
+            JSONObject dataInpFormat) throws ReflectiveOperationException {
+        Method method = SelectorHandlerUtil.class.getDeclaredMethod("fullfillSessionsVariables",
+                String.class, Map.class, JSONObject.class);
+        method.setAccessible(true);
+        return (String) method.invoke(null, whereClause, db2Input, dataInpFormat);
+    }
+
+    HashMap<String, String> callConvertToHashMap(JSONObject dataInpFormat) throws ReflectiveOperationException {
+        Method method = SelectorHandlerUtil.class.getDeclaredMethod("convertToHashMAp", JSONObject.class);
```
Comment (etendobot) [inline src-test/src/com/etendoerp/etendorx/utils/SelectorHandlerUtilBaseTest.java:109]: **Suggestion**

Consider using the 'Map' interface as the return type for better abstraction. Additionally, verify if the typo in the underlying method 'SelectorHandlerUtil.convertToHashMAp' and its corresponding reflection call 'getDeclaredMethod(\"convertToHashMAp\", ...)' should also be corrected to 'convertToHashMap' to maintain consistency.

```java
Map<String, String> callConvertToHashMap(JSONObject dataInpFormat) throws ReflectiveOperationException {
```

**Rationale**: Using the 'Map' interface instead of the concrete 'HashMap' implementation is a standard Java best practice (programming to an interface). It provides better abstraction and matches common Etendo coding patterns.

**Review by etendobot — APPROVED** [review]

**Review by etendobot — APPROVED** [review]
